### PR TITLE
[ramda] add tuple support to head and nth

### PR DESCRIPTION
**Pull request by j1r1k**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

`nth` covers just few basic cases

### CODE REVIEW

**Comment by googol** [comment]
Hi! What is the use case for this?

All the code formatting on the test file makes it hard to find what actual changes you've made there

**Comment by j1r1k** [comment]
@googol sorry about the formatting, I didn't notice it happened. I have removed the formating changes.

The benefit is that for non-empty tuple like types we actually know that return value of `head` and `nth` won't be `undefined` (for `nth` for enumerated cases at least)

**Comment by googol** [comment]
Oh of course, I see. Well that certainly can be useful. One thing that would make this even more useful would be to loosen the type constraints on the other elements of the tuple. So for head, it doesn't matter if the rest of the elements in the tuple are of type `T`, and similarly in nth only the actual nth element needs to be `T`, the others can be of any type. Do you think this could be encoded in the type?

**Comment by j1r1k** [comment]
@googol nice idea, thanks for suggestion. I have pushed another commit that adds support for heterogenous tuples

**Review by googol — APPROVED** [review]
This looks really good to me now!

File: types/ramda/index.d.ts [hunk 1]
```diff
@@ -1583,6 +1584,12 @@ declare namespace R {
         /**
          * Returns the nth element in a list.
          */
+        nth<T>(n: 0, list: [T, ...any[]]): T;
+        nth<T>(n: 1, list: [any, T, ...any[]]): T;
+        nth<T>(n: 2, list: [any, any, T, ...any[]]): T;
+        nth<T>(n: 3, list: [any, any, any, T, ...any[]]): T;
+        nth<T>(n: 4, list: [any, any, any, any, T, ...any[]]): T;
+        nth<T>(n: 5, list: [any, any, any, any, any, T, ...any[]]): T;
```
Comment (rbuckton) [inline types/ramda/index.d.ts]: I can merge this, but wouldn't the following definition allow for an arbitrary `n`?

```ts
nth<N extends number, A extends [any?, ...any[]]>(n: N, list: A): A[N];
```

Comment (googol) [reply types/ramda/index.d.ts]: I'll try this out

**Review by uniqueiniquity — CHANGES_REQUESTED** [review]
Marking as changes requested until @rbuckton's suggestion is addressed.

**Pull request closed without merging**